### PR TITLE
Allow copying in the format `cupy_array[:] = numpy_array`

### LINKED ISSUE #593: cupy_array[:] = numpy_array does not work
It would be nice if `__setitem__` worked with Numpy  arrays as right-hand side. Currently one gets the error
```
    Traceback (most recent call last):
      File "cupy/core/core.pyx", line 1575, in cupy.core.core.ndarray.__setitem__
      File "cupy/core/core.pyx", line 3043, in cupy.core.core._scatter_op
      File "cupy/core/core.pyx", line 442, in cupy.core.core.ndarray.fill
    ValueError: non-scalar numpy.ndarray cannot be used for fill
```
A straightforward implementation could first cast to a cupy array and then do the usual setitem. An optimization for same dtype and size (assuming contiguous memory) would be to just use ~~`cudaMemcpy`~~ `cupy_array.data.copy_from_host`, but that may not be very relevant.

### PR DESCRIPTION
**Pull request by pentschev**
The __setitem__ implementation from cupy.ndarray checks for an empty slice and
if the value being passed is an instance of numpy.ndarray to make a copy of it.
That can is a very useful feature in circumstances where we want to create a
copy of a numpy.ndarray without requiring mechanisms to identify the type of an
array as a CuPy array.

This resolves #593.

### CODE REVIEW

**Comment by niboshi** [comment]
@okuta How do you think?

**Comment by pentschev** [comment]
After thinking about this PR overnight, I realized there's a few improvements that could be made to make this change safer and better. Please take a look at the latest commits as well.

**Comment by pentschev** [comment]
It may be helpful also to see what we envision a use case for such a change is. Please refer to https://github.com/dask/dask-glm/pull/75/commits/e39aafd289193df56cb7e254a4aa174d3dd2dcff#diff-6fa4900cefa9ee15a64539f9fcbf1639R182.

Assume `step` is a NumPy array and `X` is a CuPy array. We then create a CuPy array `step_like` with `np.empty_like(X, shape=step.shape)`. Finally, we copy the contents via the empty slice assignment `step_like[:] = step`. In this situation, we guarantee an agnostic implementation that will convert `step` to a CuPy array from a NumPy array resulting from an algorithm that has no CuPy implementation.

Note that for this particular case, we need `np.empty_like()` to support passing an arbitrary shape. This is waiting https://github.com/numpy/numpy/pull/13046 to be merged, and after it gets merged, I have already a PR prepared to add the shape argument to CuPy as well.

**Comment by pentschev** [comment]
After a conversation with @anaruse, he said one of the concerns here was due to broadcasting. From our understanding, the issue arises when the NumPy array has a different stride, when it's a view, for example. The latest commits address that matter.

@niboshi could you confirm we understood your concerns correctly and if there are any others that we should address?

**Comment by niboshi** [comment]
Currently we intentionally disallow implicit host-to-device synchronization like this.
@okuta @asi1024 Do you have any comments?

**Comment by pentschev** [comment]
Just extending a bit on the [original reason](https://github.com/cupy/cupy/pull/2079#issue-257768572) why we want to have this special case: with the `__array_function__` protocol, it's important to have a way to permit such copies implicitly for when there different types of arrays (e.g., a NumPy and a CuPy array) coming from different sources, allowing us to target the correct implementation. Also note that this change only does _not_ allow implicit device-to-host copies, to prevent an accidental fallback to CPU, but only host-to-device.

I understand the implications this change has, so I'd be fine if we could have this in 6.0.0 as opt-in-only, through a mechanism such as an environment variable, just as NumPy does current for `__array_function__` itself.

**Comment by niboshi** [comment]
Disucssed with @asi1024 .

In this particular case (https://github.com/dask/dask-glm/pull/75/commits/e39aafd289193df56cb7e254a4aa174d3dd2dcff#diff-6fa4900cefa9ee15a64539f9fcbf1639R182), it seems that after #2165 is merged, the code should work because `cupy.linalg.lstsq` will return `cupy.ndarray` as `step`. Even without #2165, we're considering adding "NumPy-fallback" feature that should make the code work in this case. In this feature, even non-implemented functions (`cupy.linalg.lstsq`) should work by accepting and returning cupy.ndarrays and falling back to numpy's corresponding functions.

As for mixed-array-type case you mentioned above, could you elaborate why you need such cases supported?

**Comment by pentschev** [comment]
Thanks @niboshi and @asi1024 for checking on this, and for letting me know about #2165. This would certainly solve that particular case from dask-glm.

For the mixed arrays, consider a slightly different example in the same dask-glm PR, https://github.com/dask/dask-glm/commit/e39aafd289193df56cb7e254a4aa174d3dd2dcff#diff-6fa4900cefa9ee15a64539f9fcbf1639R240. This function receives a `beta` which is the output of some fortran code in scipy, which isn't easy to be ported either to pure NumPy nor CuPy, so we're bound to get this output as a NumPy array. However, `X` could be, depending on the user input, either a NumPy array, or a CuPy array. It's this last case that we're interested in, we want to have `beta` copied to a CuPy array, so the remaining computation can be done in the GPU.

I hope this clarifies a bit the sort of use case we have for this.

As for the "NumPy-fallback" feature you mentioned, would it work to wrap a third-party functions (like SciPy) as well? For example, could we pass CuPy arrays to a pure-CPU implementation in SciPy and ensure we get back a CuPy array? I can't think of a way to do that, but if we could, that would be great and probably solve this as well.

**Comment by niboshi** [comment]
>For the mixed arrays, consider a slightly different example in the same dask-glm PR, https://github.com/dask/dask-glm/commit/e39aafd289193df56cb7e254a4aa174d3dd2dcff#diff-6fa4900cefa9ee15a64539f9fcbf1639R240. This function receives a `beta` which is the output of some fortran code in scipy, which isn't easy to be ported either to pure NumPy nor CuPy, so we're bound to get this output as a NumPy array. However, `X` could be, depending on the user input, either a NumPy array, or a CuPy array. It's this last case that we're interested in, we want to have `beta` copied to a CuPy array, so the remaining computation can be done in the GPU.

In this case, as you know which argument *may* be of differerent types, I think you can write `np.asarray(beta)` before passing to `func` (I assume `np` may be `numpy` or `cupy`). Do you think that works?

>As for the "NumPy-fallback" feature you mentioned, would it work to wrap a third-party functions (like SciPy) as well? For example, could we pass CuPy arrays to a pure-CPU implementation in SciPy and ensure we get back a CuPy array? I can't think of a way to do that, but if we could, that would be great and probably solve this as well.

I don't think such cases will be supported either (@asi1024 How do you think?).
Even so, I think `np.asarray` method above can be used in any case.

**Comment by pentschev** [comment]
> In this case, as you know which argument _may_ be of differerent types, I think you can write `np.asarray(beta)` before passing to `func` (I assume `np` may be `numpy` or `cupy`). Do you think that works?

I don't think that works, we would have to assume that we know `beta` needs to be a CuPy array, but that isn't always true, as it depends on the type of `X` array.

**Comment by pentschev** [comment]
Maybe my last comment wasn't very clear, when I say `beta` needs to be a CuPy array, I mean that `beta` needs to be converted to a CuPy array, depending on whether `X` is also a CuPy array.

**Comment by niboshi** [comment]
My assumption was `np` may be either `numpy` or `cupy` that matches the type of `X`.

If this assumption is incorrect and `np` is always `numpy`, you can use `xp = cupy.get_array_module(X)` and `xp.asarray`.

**Comment by pentschev** [comment]
> My assumption was `np` may be either `numpy` or `cupy` that matches the type of `X`.
> 
> If this assumption is incorrect and `np` is always `numpy`, you can use `xp = cupy.get_array_module(X)` and `xp.asarray`.

Sure, this would work. But the point of having `__array_function__` in place is to avoid the need to handle explicitly other libraries (and even to let internals know about their existence), and rely solely on NumPy to do the whole work, no matter what the array type.

We understand the potential harm for allowing implicit copies, so having this special copy `cupy_array[:] = numpy_array` would allow libraries such as dask-glm to be agnostic of CuPy and at the same time prevent unnecessary copies elsewhere, other than a few places where this is absolutely necessary, since during implementation we know of this restriction when using SciPy, or any other CPU-only library.

**Comment by niboshi** [comment]
Ah, OK, I missed the existence of other libraries.

**Comment by pentschev** [comment]
Hi @niboshi, just wanted to check in with you, does my previous example make sense? What else do we need to move forward with this PR? I'm happy to clarify and discuss this further if needed.

**Comment by pentschev** [comment]
I made this change opt-in only by introducing `CUPY_EXPERIMENTAL_SLICE_COPY` environment variable, I also allowed tests to skip when disabled or unset.

I think it would be nice to have this tested as well, which means we would have to setup one of the tests with that variable or add a new one. On Dask we did exactly that, setup only one of the tests to enable environment variables, while the others do not change, thus incurring in no extra build time.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1183,7 +1183,27 @@ cdef class ndarray:
...
+                    ptr = ctypes.c_void_p(tmp.__array_interface__['data'][0])
+                stream_ptr = stream_module.get_current_stream_ptr()
+                if stream_ptr == 0:
+                    self.data.copy_from_host(ptr, self.nbytes)
+                else:
+                    self.data.copy_from_host_async(ptr, self.nbytes)
+            else:
+                raise ValueError(
+                    "copying a numpy.ndarray to a cupy.ndarray by empty slice "
+                    "assignment must ensure arrays exact same shape and dtype")
```
Comment (niboshi) [inline cupy/core/core.pyx]: Please use single quotes for consistency.

Comment (pentschev) [reply cupy/core/core.pyx]: Thanks for noticing that, thought that flake8 would catch this, but I realized it doesn't support that, at least without third-party plugins. Just pushed a fix for this.

Comment (niboshi) [reply cupy/core/core.pyx]: Thanks!

**Comment by niboshi** [comment]
Thank you for the fix.
Jenkins, test this please.

**Comment by pentschev** [comment]
Looks like the failure is in chainer-doc, probably not related to this PR.

**Comment by niboshi** [comment]
Looks so. Retrying.
Jenkins, test this please

**Comment by niboshi** [comment]
Thanks. LGTM!

**Comment by pentschev** [comment]
Awesome! Thanks @niboshi for the review and merging!

**Comment by pentschev** [comment]
By the way, since this now requires explicit opt-in, do you think it could be backported to the next CuPy 6 release?

**Comment by niboshi** [comment]
Yes, sure. Thank you for pointing out!

**Pull request merged**